core/src/price_reporter/exchange/kraken.rs

```diff
@@ -0,0 +1,171 @@
+//! Defines an abstraction over a Kraken WS connection
+
+use std::{
+    collections::HashSet,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use async_trait::async_trait;
+use futures_util::{Sink, SinkExt, Stream, StreamExt};
+use lazy_static::lazy_static;
+use serde_json::json;
+use tracing::log;
+use tungstenite::{Error as WsError, Message};
+use url::Url;
+
+use crate::price_reporter::{
+    errors::ExchangeConnectionError,
+    exchange::{connection::ws_connect, Exchange},
+    reporter::Price,
+    tokens::Token,
+    worker::PriceReporterManagerConfig,
+};
+
+use super::{
+    connection::{parse_json_field_array, parse_json_from_message, ExchangeConnection},
+    InitializablePriceStream,
+};
+
+// -------------
+// | Constants |
+// -------------
+
+/// The name of the events field in a Kraken WS message
+const KRAKEN_EVENT: &str = "event";
+/// The index of the price data in a Kraken WS message
+const KRAKEN_PRICE_DATA_INDEX: usize = 1;
+/// The index of the bid price in a Kraken WS message's price data
+const KRAKEN_BID_PRICE_INDEX: usize = 0;
+/// The index of the ask price in a Kraken WS message's price data
+const KRAKEN_ASK_PRICE_INDEX: usize = 1;
+/// The timestamp of the price report from kraken
+const KRAKEN_PRICE_REPORT_TIMESTAMP_INDEX: usize = 2;
+
+lazy_static! {
+    static ref KRAKEN_MSG_IGNORE_LIST: HashSet<String> = {
+        let mut set = HashSet::new();
+
+        set.insert(String::from("systemStatus"));
+        set.insert(String::from("subscriptionStatus"));
+        set.insert(String::from("heartbeat"));
+        set
+    };
+}
+
+// -----------------------------
+// | Connection Implementation |
+// -----------------------------
+
+/// The message handler for Exchange::Kraken.
+pub struct KrakenConnection {
+    /// The underlying price stream
+    price_stream: Box<dyn Stream<Item = Price> + Unpin>,
+    /// The underlying write stream of the websocket
+    write_stream: Box<dyn Sink<Message, Error = WsError>>,
+}
+
+impl KrakenConnection {
+    /// Get the URL for the Kraken websocket endpoint
+    fn websocket_url() -> Url {
+        String::from("wss://ws.kraken.com")
+            .parse()
+            .expect("Failed to parse Kraken websocket URL")
+    }
+
+    /// Parse a price report from a Kraken websocket message
+    fn midpoint_from_ws_message(
```

Comment:
I think "What can we fit in a trait to avoid redudancy" is a bad way to think about it. Instead, the `ExchangeConnection` trait focuses on a minimal consumable interface and does not attempt to paper over odd ownership and stateful differences that come up across exchanges. Plus we can implement `ExchangeConnection` for Uniswap as well, which gives a clean interface in higher layers.

I'll add a more peripheral point: I think DRY is a bad principal in general, and especially bad for rust programming. DRY hurts readability, discoverability, extensibility at the benefit of reduced redundancy. Rust gives us endless ways to show off and build bone-DRY interfaces. The end result of the two is an over-abstracted mess IMO